src/dataflow/src/render/sources.rs

```diff
@@ -514,3 +515,79 @@ where
+fn flatten_results<G, T>(
+    key_envelope: KeyEnvelope,
+    results: timely::dataflow::Stream<Child<G, T>, DecodeResult>,
+) -> (
+    timely::dataflow::Stream<Child<G, T>, Row>,
+    timely::dataflow::Stream<Child<G, T>, DataflowError>,
+)
+where
+    G: ScopeParent,
+    T: Refines<<G as ScopeParent>::Timestamp>,
+{
+    match key_envelope {
+        KeyEnvelope::None => results
+            .flat_map(|DecodeResult { key: _, value, .. }| value)
+            .ok_err(std::convert::identity),
+        KeyEnvelope::Flattened | KeyEnvelope::LegacyUpsert => results
+            .flat_map(flatten_key_value)
+            .map(|(key, value)| {
+                match (key, value) {
+                    (Ok(mut key), Ok(value)) => {
+                        key.extend_by_row(&value);
+                        Ok(key)
+                    }
+                    // always prioritize the value error if either or both have an error
+                    (_, Err(e)) => Err(e),
+                    (Err(e), _) => Err(e),
+                }
+            })
+            .ok_err(std::convert::identity),
+        KeyEnvelope::Named(_) => results
+            .flat_map(flatten_key_value)
+            .map(|(key, value)| match (key, value) {
+                (Ok(mut key), Ok(value)) => {
+                    // Named semantics rename a key that is a single column, and encode a
+                    // multi-column field as a struct with that name
+                    match key.iter().count() {
+                        1 => {
+                            key.extend_by_row(&value);
+                            Ok(key)
+                        }
+                        _ => {
+                            let mut new_row = Row::default();
+                            new_row.push_list(key.iter());
+                            new_row.extend_by_row(&value);
+                            Ok(new_row)
+                        }
+                    }
+                }
+                // always prioritize the value error if either or both have an error
+                (_, Err(e)) => Err(e),
+                (Err(e), _) => Err(e),
+            })
+            .ok_err(std::convert::identity),
+    }
+}
+
+/// Handle possibly missing key or value portions of messages
+fn flatten_key_value(
+    result: DecodeResult,
+) -> Option<(Result<Row, DataflowError>, Result<Row, DataflowError>)> {
+    let DecodeResult { key, value, .. } = result;
+    match (key, value) {
+        (Some(key), Some(value)) => Some((key, value)),
+        (None, None) => None,
+        // the next map where this is used in `flatten_results` prioritizes the value error, so the
+        // key doesn't matter here
+        _ => Some((
+            Ok(Row::default()),
+            Err(DataflowError::DecodeError(DecodeError::Text(
+                "Key and/or Value are not present for message".to_string(),
+            ))),
+        )),
```

Comment:
Indeed, this will work correctly, but why not make it return `Result<Option<(Row, Row)>, DataflowError>` to avoid this somewhat confusing logic?

Reply:
oh good point.